service/history/workflow/mutable_state_impl.go

```diff
@@ -6252,3 +6259,259 @@ func (ms *MutableStateImpl) RefreshExpirationTimeoutTask(ctx context.Context) er
+	})
+
+	applyUpdatesToSubStateMachine(ms, ms.pendingRequestCancelInfoIDs, ms.updateRequestCancelInfos, updatedRequestCancelInfos, currentVersionedTransition, nil)
+	applyUpdatesToSubStateMachine(ms, ms.pendingSignalInfoIDs, ms.updateSignalInfos, updatedSignalInfos, currentVersionedTransition, nil)
+}
+
+func (ms *MutableStateImpl) applyUpdatesToStateMachineNodes(
+	nodeMutations []*persistencespb.WorkflowMutableStateMutation_StateMachineNodeMutation,
+) error {
+	root := ms.HSM()
+	for _, nodeMutation := range nodeMutations {
+		incomingPath := []hsm.Key{}
+		for _, p := range nodeMutation.Path.Path {
+			incomingPath = append(incomingPath, hsm.Key{Type: p.Type, ID: p.Id})
+		}
+		currentNode, err := root.Child(incomingPath)
+		if err != nil {
+			if !errors.Is(err, hsm.ErrStateMachineNotFound) {
+				return err
+			}
+			parent := incomingPath[:len(incomingPath)-1]
+			if len(parent) == 0 {
+				parent = root.Path()
+			}
+			parentNode, err := root.Child(parent)
+			if err != nil {
+				// we don't have enough information to recreate all parents
+				return err
+			}
+			currentNode, err = parentNode.AddChild(incomingPath[len(incomingPath)-1], nodeMutation.Data)
+			if err != nil {
+				return err
+			}
+		}
+		internalNode := currentNode.InternalRepr()
+		if CompareVersionedTransition(nodeMutation.LastUpdateVersionedTransition, internalNode.LastUpdateVersionedTransition) <= 0 {
+			continue
+		}
+		internalNode.Data = nodeMutation.Data
+		internalNode.InitialVersionedTransition = nodeMutation.InitialVersionedTransition
+		internalNode.LastUpdateVersionedTransition = nodeMutation.LastUpdateVersionedTransition
+		internalNode.TransitionCount++
+	}
+	ms.mustInitHSM()
+	return nil
+}
+
+func (ms *MutableStateImpl) applySignalRequestedIds(signalRequestedIds []string, incomingExecutionInfo *persistencespb.WorkflowExecutionInfo) {
+	incomingVersionedTransition := incomingExecutionInfo.TransitionHistory[len(incomingExecutionInfo.TransitionHistory)-1]
+	if CompareVersionedTransition(incomingVersionedTransition, ms.executionInfo.SignalRequestIdsLastUpdateVersionedTransition) > 0 {
+		for requestID := range ms.pendingSignalRequestedIDs {
+			ms.approximateSize -= len(requestID)
+		}
+		ms.pendingSignalRequestedIDs = convert.StringSliceToSet(signalRequestedIds)
+		for requestID := range ms.pendingSignalRequestedIDs {
+			ms.approximateSize += len(requestID)
+		}
+		ms.executionInfo.SignalRequestIdsLastUpdateVersionedTransition = incomingVersionedTransition
+	}
+}
+
+func applyUpdatesToSubStateMachine[K comparable, V lastUpdatedStateTransitionGetter](
+	ms *MutableStateImpl,
+	pendingInfos map[K]V,
+	updateInfos map[K]V,
+	updatedSubStateMachine map[K]V,
+	versionedTransition *persistencespb.VersionedTransition,
+	fn func(current, incoming V),
+) {
+	for key, updated := range updatedSubStateMachine {
+		if CompareVersionedTransition(updated.GetLastUpdateVersionedTransition(), versionedTransition) <= 0 {
+			continue
+		}
```

Comment:
hmm I don't think this holds. Individual sub state machine can have a lower versioned transition as long as the overall mutable state versioned transition is higher. Apply mutation/snapshot should update as long as current lastUpdateVersionedTransition != incoming.lastUpdateVersionedTransition.